### PR TITLE
Domain: fix form

### PR DESCRIPTION
**Pull request by stsewd**
- Allow to save the same domain with canonical=true
- Don't save domains like `domain.com/foo`

### CODE REVIEW

File: readthedocs/rtd_tests/tests/test_domains.py [hunk 1]
```diff
@@ -99,6 +99,15 @@ def test_public_domain_not_allowed(self):
...
         )
 
+    def test_domain_with_path(self):
+        form = DomainForm(
+            {'domain': 'domain.com/foo/bar'},
+            project=self.project,
+        )
+        self.assertTrue(form.is_valid())
+        domain = form.save()
+        self.assertEqual(domain.domain, 'domain.com')
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_domains.py:109]: Not sure if we should raise an error instead... But looking at the code this is what we wanted in the beginning?

Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_domains.py:109]: Actually that's what we are doing with domains like `https://foo.com/foo/bar`

File: readthedocs/projects/forms.py [hunk 2]
```diff
@@ -621,11 +621,19 @@ def clean_project(self):
...
-        else:
-            domain_string = parsed.path
+        domain = self.cleaned_data['domain']
+        parsed = urlparse(domain)
+
+        # Force the scheme to have a valid netloc.
+        if not parsed.scheme:
+            parsed = urlparse(f'https://{domain}')
+
+        if not parsed.netloc:
```
Comment (agjohnson) [inline readthedocs/projects/forms.py:631]: This is much better than nothing, but it isn't a great check:

```
In [13]: parse.urlparse('https://*****.laksdhg.foo')
Out[13]: ParseResult(scheme='https', netloc='*****.laksdhg.foo', path='', params='', query='', fragment='')
```

File: readthedocs/projects/forms.py [hunk 3]
```diff
@@ -638,8 +646,14 @@ def clean_domain(self):
...
+        pk = self.instance.pk
+        has_canonical_domain = (
+            Domain.objects
+            .filter(project=self.project, canonical=True)
+            .exclude(pk=pk)
+            .exists()
+        )
+        if canonical and has_canonical_domain:
             raise forms.ValidationError(
                 _('Only 1 Domain can be canonical at a time.'),
```
Comment (agjohnson) [inline readthedocs/projects/forms.py:658]: Might as well also fix this error message, domain is not a proper noun.

```suggestion
                _('Only one domain can be canonical at a time.'),
```

File: readthedocs/rtd_tests/tests/test_domains.py (hunk 1 above)
Comment (agjohnson) [inline readthedocs/rtd_tests/tests/test_domains.py:109]: It might be worth adding some test cases with more garbage data as well too

**Review by agjohnson — COMMENTED** [review]
Looks good, i only had some minor improvements to suggest to make this more defensive.

File: readthedocs/projects/forms.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/projects/forms.py:631]: Looks like something the domain validator should be catching https://github.com/readthedocs/readthedocs.org/blob/b74fe8d5aab8421fb194d7b8cbbb42a4e54f2569/readthedocs/projects/validators.py#L15-L24

But isn't... Somehow that regex is allowing that value, I took at look at the new version of that validator https://github.com/django/django/pull/7300/files but the result is the same. Also, I think we shouldn't allow ip values here

File: readthedocs/rtd_tests/tests/test_domains.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_domains.py:107]: I have added some, but mostly aren't being caught by https://github.com/readthedocs/readthedocs.org/pull/7502/files#r493113673

File: readthedocs/rtd_tests/tests/test_domains.py [hunk 4]
```diff
@@ -99,6 +97,28 @@ def test_public_domain_not_allowed(self):
...
+
+    def test_invalid_domains(self):
+        invalid = [
+            'python..org',
+            # '****.foo.com', current validator says this is valid :shrug:
```
Comment (humitos) [inline readthedocs/rtd_tests/tests/test_domains.py:112]:
```suggestion
            # FIXME: '****.foo.com', current validator says this is valid :shrug:
```

Can we add a test for domains with path also, like `domain.com/path/to/something`?

Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_domains.py]: The above test is testing that

Comment (humitos) [reply readthedocs/rtd_tests/tests/test_domains.py]: :upside_down_face:

**Review by ericholscher — APPROVED** [review]
This looks good to me as a start at least. 👍

**Pull request merged**